### ISSUE TITLE
Document why conbench uses history to measure variability

### ISSUE THREAD

**Issue by jonkeane** [conbench/conbench#583]
This is come up in a few places and is a bit of an unintuitive trick, but one that can have nice benefits. We should document _why_ we do this a bit more so that folks who seek to understand what we're using from history and why.

**Comment by jgehrcke** [comment]
Thanks for this ticket here. Yes, I'd love to read / hear about all those good reasons for looking into the historical evolution of data in order to extract useful knowledge from there. I trust that it's possible, and also valuable.

In my current world-view the relevance of this is however limited.

My current thinking on priorities towards good regression detection:


1. **Create good raw data**. I believe that this can get us almost all the way towards decent regression detection with trivial data analysis. Reduce noise, analyze the nature of noise.
2. **Quantification of distribution (variability/noise) mainly from _now_**. This relates to (1), but deserves its own priority item. Multisampling with N>>1 is useful, but it’s also expensive. Interesting question: how big must N be so that we do _not_ need to go into the history, but can compare two multisample results directly? Having a large (enough) N for actually constant conditions is better than trying to get some variability data from the past where conditions were not-so-constant (but are assumed to be!).
3. **Work with user feedback on candidates**. This tries to pre-detect certain events via simple heuristics. Those events then need user feedback: e.g., regression, yes or no? The candidate creation plus human feedback itself creates a historical data set of potentially great value ('learning from humans', relation to AI/ML!)
3. **Extraction of variability/noise/patterns from the past**. Yes, historical evolution can give clues that might help with regression detection. But it’s tough to make this reliable, smart. I think if 1 and 2 are done well, then we get great results for 95 % of the interesting cases. For the last few percent we can try to go crazy.

I think that 1&2 will fundamentally be the most fruitful areas to focus on. Good return-of-investment to be expected. I also think that the limitations of 1&2 should be well-explored and understood before investing too much into 3 and 4.

I nevertheless of course think 3 and 4 are super interesting (and somewhat equally important).

Of course I am super open to my thinking to change over time. Just wanted to share my current mental model.

**Comment by alistaire47** [comment]
This is making me think of a power calculation for a T-test—using history gets you greater statistical power at the expense of the assumption that the data is from the same distribution (which is may or may not be). Not saying we should start doing power calculations, but maybe it's a useful framing for figuring out when we can say something useful and when we can't.

**Comment by jgehrcke** [comment]
> using history gets you greater statistical power at the expense of the assumption that the data is from the same distribution (which is may or may not be).

Thank you @alistaire47 for this thoughtful statement about this fundamental challenge here!

**Comment by jonkeane** [comment]
> Power calculation [...] using history gets you greater statistical power at the expense of the assumption that the data is from the same distribution (which is may or may not be).

Absolutely agreed this is a great point, and something highlighting (at least in a detailed deep-dive type doc) will help elucidate. Strong agree that we probably don't want to do a traditional power test in the same way that we use Z-scores for their properties of measuring how compatible an observation is with a distribution, but we don't use the traditional cut off of 2/1.96 since that would mean that 5% of the time (for each benchmark!) when there is no regression we would alert saying that there is one. But using a higher threshold can balance that false positive rate for the application we're using it.

**Comment by alistaire47** [comment]
> we don't use the traditional cut off of 2/1.96 since that would mean that 5% of the time (for each benchmark!) when there is no regression we would alert saying that there is one. But using a higher threshold can balance that false positive rate for the application we're using it.

If my math is right, using a threshold of 5 sigma would correspond to a FP rate of 0.00573% which is roughly 3 FP / 5,000,000 instead of 1/20 for 1.96. (Assuming we're testing means of enough observations so the CLT holds and they're normally distributed; bimodality and low iterations may bork this.) I think the stats framework largely holds up here, just the magic numbers of academic publishing are not what we want.

**Comment by jonkeane** [comment]
> If my math is right, using a threshold of 5 sigma would correspond to a FP rate of 0.00573% which is roughly 3 FP / 5,000,000 instead of 1/20 for 1.96. (Assuming we're testing means of enough observations so the CLT holds and they're normally distributed; bimodality and low iterations may bork this.)

Yup totally. What is the right number is an empirical question we can and should measure + decide (we have all the data, for Arrow at least!). My original work on this (now > 1 year ago) is the ideal threshold is ~9 for a decent rate of true positives and low false positives. We ended up settling on 5 because it was halfway between 9 and that magic ac-publishing number (there was concern at the time that 9 was too liberal and would basically _never_ flag a regression). We can change that threshold whenever we want (e.g. I wouldn't be opposed to upping it to 9 today) and we _definitely_ should circle back and do a real analysis on this now that we have going on a few years of data to tell what behavior is like.

> I think the stats framework largely holds up here, just the magic numbers of academic publishing are not what we want.

Yup, totally. Like you mention above: we do play a little fast and loose with some assumptions of normality that would also bite us if we were using these in a more academic + scientific manner.
